ray-operator/controllers/ray/expectations/active_expectation.go

```diff
@@ -0,0 +1,382 @@
+		return fmt.Errorf("fail to expect %s/%s for action %d, %s: %s", kind, name, Update, updatedResourceVersion, err)
+	}
+
+	return nil
+}
+
+func (ae *ActiveExpectations) IsSatisfied(key string) (satisfied bool, err error) {
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return false, fmt.Errorf("fail to get expectation for active expectations %s when check satisfication: %s", key, err)
+	}
+
+	if !exist {
+		return true, nil
+	}
+
+	defer func() {
+		if satisfied {
+			ae.subjects.Delete(expectation)
+		}
+	}()
+
+	satisfied, err = expectation.(*ActiveExpectation).isSatisfied()
+	if err != nil {
+		return false, err
+	}
+
+	return
+}
+
+func (ae *ActiveExpectations) DeleteItem(key string, kind ExpectedResourceType, name string) error {
+	if _, exist := ResourceInitializers[kind]; !exist {
+		panic(fmt.Sprintf("kind %s is not supported for Active Expectation", kind))
+	}
+
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get expectation for active expectations %s when deleting name %s: %s", key, name, err)
+	}
+
+	if !exist {
+		return nil
+	}
+
+	item := expectation.(*ActiveExpectation)
+	if err := item.delete(string(kind), name); err != nil {
+		return fmt.Errorf("fail to delete %s/%s for key %s: %s", kind, name, key, err)
+	}
+
+	if len(item.items.List()) == 0 {
+		ae.subjects.Delete(expectation)
+	}
+
+	return nil
+}
+
+func (ae *ActiveExpectations) Delete(key string) error {
+	expectation, exist, err := ae.subjects.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get expectation for active expectations %s when deleting: %s", key, err)
+	}
+
+	if !exist {
+		return nil
+	}
+
+	err = ae.subjects.Delete(expectation)
+	if err != nil {
+		return fmt.Errorf("fail to do delete expectation for active expectations %s when deleting: %s", key, err)
+	}
+
+	return nil
+}
+
+func ActiveExpectationItemKeyFunc(object interface{}) (string, error) {
+	expectationItem, ok := object.(*ActiveExpectationItem)
+	if !ok {
+		return "", fmt.Errorf("fail to convert to active expectation item")
+	}
+	return expectationItem.Key, nil
+}
+
+func NewActiveExpectation(client client.Client, key, namespace string) *ActiveExpectation {
+	return &ActiveExpectation{
+		Client:          client,
+		namespace:       namespace,
+		key:             key,
+		items:           cache.NewStore(ActiveExpectationItemKeyFunc),
+		recordTimestamp: time.Now(),
+	}
+}
+
+type ActiveExpectation struct {
+	client.Client
+	namespace       string
+	key             string
+	items           cache.Store
+	recordTimestamp time.Time
+}
+
+func (ae *ActiveExpectation) expectCreateOrDelete(kind ExpectedResourceType, name string, action ActiveExpectationAction) error {
+
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get active expectation item for %s when expecting: %s", key, err)
+	}
+
+	ae.recordTimestamp = time.Now()
```

Comment:
Should we use a mutex for updating the `recordTimestamp`?

Reply:
> Should we use a mutex for updating the `recordTimestamp`?

Thanks for your review. 😺

The use of a mutex for `ActiveExpectation`'s `recordTimestamp` depends on the context of the `ActiveExpectations`. Currently, it is only employed in the controller's `Reconcile` func. Within the controller, multiple workers reconcile concurrently, but only one worker handles a given `reconcile.Request` at any given time.

Whether the `recordTimestamp` is used with a mutex in `ActiveExpectation` depends on its usage context. It is currently only used within the `Reconcile` func in the controller, where multiple workers are running in parallel. However, for the same `ReconcileRequest`, only a single worker handles it at any given time. 

As a result, only one goroutine handles the `ActiveExpectation`s associated with the same `RayCluster`, ensuring there won't be any concurrency issues with reading and writing.

However, this can cause issues if `ActiveExpectations` are used externally by something like an `EventHandler`. But we aren't seeing this use case currently.

Reply:
👍